test/env/data.go

```diff
@@ -179,3 +184,79 @@ func BundleHasSyncedAllNamespaces(ctx context.Context, cl client.Client, name, e
 
 	return true
 }
+
+// CheckBundleSynced returns nil if the given Bundle has synced the expected data to the given
+// namespace, or else returns a descriptive error if that's not the case.
+// - Skips Namespaces that are Terminating since targets are not synced there.
+// - Ensures the Bundle status has been updated with the appropriate target.
+// - Ensures the Bundle has the correct status condition with the same ObservedGeneration as the current Generation.
+func CheckBundleSynced(ctx context.Context, cl client.Client, bundleName string, namespace string, expectedData string) error {
+	var bundle trustapi.Bundle
+	Expect(cl.Get(ctx, client.ObjectKey{Name: bundleName}, &bundle)).NotTo(HaveOccurred())
+
+	var configMap corev1.ConfigMap
+	if err := cl.Get(ctx, client.ObjectKey{Namespace: namespace, Name: bundle.Name}, &configMap); err != nil {
+		return fmt.Errorf("failed to get configMap %s/%s when checking bundle sync: %w", namespace, bundle.Name, err)
+	}
+
+	gotData := configMap.Data[bundle.Spec.Target.ConfigMap.Key]
+
+	if gotData != expectedData {
+		// TODO: also print expected and got data, but don't just dump the raw PEM values - maybe parse certs and transform into a user friendly representation for easier visual inspection
+		return fmt.Errorf("configMap %s/%s didn't have expected value", namespace, bundle.Name)
+	}
+
+	if bundle.Status.Target == nil {
+		return fmt.Errorf("bundle status target was nil")
+	}
+
+	if !apiequality.Semantic.DeepEqual(*bundle.Status.Target, bundle.Spec.Target) {
+		return fmt.Errorf("bundle status target didn't match expected status target")
+	}
+
+	for _, condition := range bundle.Status.Conditions {
+		if condition.Status == corev1.ConditionTrue && bundle.Generation == condition.ObservedGeneration {
+			return nil
+		}
+	}
+
+	return fmt.Errorf("couldn't find a success condition on bundle status with expected observedGeneration %d", bundle.Generation)
+}
+
+// bundleHasSyncedAllNamespacesError calls CheckBundleSynced for all namespaces and returns an error if any of them failed
+func bundleHasSyncedAllNamespacesError(ctx context.Context, cl client.Client, name, expectedData string) error {
+	var namespaceList corev1.NamespaceList
+	if err := cl.List(ctx, &namespaceList); err != nil {
+		return fmt.Errorf("failed to list namespaces: %w", err)
+	}
+
+	var errs []string
+
+	for _, namespace := range namespaceList.Items {
+		// Skip terminating namespaces since Bundle won't be synced there
+		if namespace.Status.Phase == corev1.NamespaceTerminating {
+			continue
+		}
+
+		if err := CheckBundleSynced(ctx, cl, name, namespace.Name, expectedData); err != nil {
+			errs = append(errs, fmt.Errorf("namespace %q: %w", namespace.Name, err).Error())
+		}
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("bundle has not synced all namespaces. namespaces with errors: %s", strings.Join(errs, " | "))
```

Comment:
instead of joining the strings, can we use `utilerrors "k8s.io/apimachinery/pkg/util/errors"` here to create a combined error & wrap that error using `fmt.Errorf`?

Reply:
I don't think that's functionally different to the current implementation. It might be a little more "future proof" but I'm not really anticipating that this function would need massively changing in the future.

I don't know what we'd really gain from wrapping these errors, because we're not _realistically_ going to be digging into their context, IMO.

I think I disagree with the suggestion, but I'll do it anyway!

Reply:
Interestingly, something related to this is coming in go 1.20!

https://tip.golang.org/doc/go1.20#errors